[WIP] Concepts doc updates (#4219)
circleci/circleci-docs: jekyll/_cci2/concepts.md

{% include snippets/pipelines-benefits.adoc %}

## Orbs
Orbs are reusable snippets of code that help automate repeated processes, speed up project setup, and make it easy to integrate with third-party tools. See [Using Orbs]({{ site.baseurl }}/2.0/using-orbs/) for details about how to use orbs in your config and an introduction to orb design. Visit the [Orbs Registry](https://circleci.com/orbs/registry/) to search for orbs to help simplify your config.

## Config

CircleCI believes in *configuration as code*. The entire pipeline process, from build to deploy, is orchestrated through a single file called `config.yml`.  The `config.yml` file is located in a folder called `.circleci` at the root of your project. CircleCI uses the YAML syntax for config, see the [Writing YAML]({{ site.baseurl }}/2.0/writing-yaml/) document for basics.

```sh
├── .circleci
│   ├── config.yml
├── README
└── all-your-projects-other-files-and-folders
```

`config.yml` is a powerful yaml file that defines the entire pipeline for your project. For a full overview of the various keys that can be used see the [Configuration Reference]({{ site.baseurl }}/2.0/configuration-reference/). 

> Put very simply, you will define [workflows](#workflows) that orchestrate a series of [jobs](#jobs). Each job will contain a number of [steps](#steps) to run commands and shell scripts to do the work required for your project. Each job runs in an independent [executor](#executors-and-images) (container or virtual machine). [Caches](#caches-workspaces-and-artifacts) are available to optimize and speed up pipelines, and [workspaces/artifacts](#caches-workspaces-and-artifacts) can be used to share data across your pipeline.

Reviewer: I'm wondering if this is worth breaking down into an un/ordered list? Thoughts?

Author: Yes, I think that would work well. I'll try it.

The following image uses an [example Java application](https://github.com/CircleCI-Public/circleci-demo-java-spring/tree/2.1-config) to show the various config elements (**Note:** The `test` job used in the example has been removed fro the image for the purposes of keeping the image a reasonable size):

![config elements]({{ site.baseurl }}/assets/img/docs/config-elements.png)

> And now the [same example using the Maven orb](https://github.com/CircleCI-Public/circleci-demo-java-spring/tree/2.1-orbs-config) illustrates how orbs can be used to simplify your project configuration:

Reviewer: Nice. Good idea.

![config elements orbs]({{ site.baseurl }}/assets/img/docs/config-elements-orbs.png)

### Jobs

Jobs are the building blocks of your config. Jobs are collections of [steps](#steps), which run commands/scripts as required. Each job must declare an executor that is either `docker`, `machine`, `windows` or `macos`. `machine` includes a [default image](https://circleci.com/docs/2.0/executor-intro/#machine) if not specified, for `docker` you must [specify an image](https://circleci.com/docs/2.0/executor-intro/#docker) to use for the primary container, for `macos` you must specify an [Xcode version](https://circleci.com/docs/2.0/executor-intro/#macos), and for `windows` you must use the [Windows orb](https://circleci.com/docs/2.0/executor-intro/#windows).

![job illustration]( {{ site.baseurl }}/assets/img/docs/job.png)

#### Executors and Images
{:.no_toc}

> Each separate job defined within your config will run in a unique executor. An executor can be a docker container or a virtual machine running Linux, windows, or macOS. Note, macOS is not currently available on self-hosted installations of CircleCI Server.

Reviewer: ```suggestion
Each separate job defined within your config will run in a unique executor. An executor can be a docker container or a virtual machine running Linux, Windows, or MacOS. Note, macOS is not currently available on self-hosted installations of CircleCI Server.
```

* TOC 
{:toc}

## Add Projects

On the "Add Projects" page, you can either _Set Up_ any project that you are
the owner of on your VCS, or, _Follow_ any project in your organization to gain
access to its pipelines and to subscribe to [email notifications]({{
site.baseurl }}/2.0/notifications/) for the project's status.

{:.tab.addprojectpage.Cloud}
![header]({{ site.baseurl }}/assets/img/docs/CircleCI-2.0-setup-project-circle101_cloud.png)

{:.tab.addprojectpage.Server}
![header]({{ site.baseurl }}/assets/img/docs/CircleCI-2.0-setup-project-circle101.png)

## User Types

The *Project Administrator* is the user who adds a GitHub or Bitbucket
repository to CircleCI as a Project. A *User* is an individual user within an
org. A CircleCI user is anyone who can log in to the CircleCI platform with a
username and password. Users must be added to a [GitHub or Bitbucket org]({{
site.baseurl }}/2.0/gh-bb-integration/) to view or follow associated CircleCI
projects.  Users may not view project data that is stored in environment variables.

## Config

> CircleCI believes in *configuration as code*. The entire pipeline process, from build to deploy, is orchestrated through a single file called `config.yml`.  The `config.yml` file is located in a folder called `.circleci` at the top of your project. CircleCI uses the YAML syntax for config, see the [Writing YAML]({{ site.baseurl }}/2.0/writing-yaml/) document for basics.

Reviewer: I would recommend `root of your project` instead of `top`.

Author: Oh yeah good call 👍 :-)

```sh
├── .circleci
│   ├── config.yml
├── README
└── all-your-projects-other-files-and-folders
```

`config.yml` is a powerful yaml file that defines the entire pipeline for your project. For a full overview of the various keys that can be used see the [Configuration Reference]({{ site.baseurl }}/2.0/configuration-reference/). 

> Put very simply, you will define [workflows](#workflows) that orchestrate a series of [jobs](#jobs). Each job will contain a number of [steps](#steps) to run commands and shell scripts to do the work required for your project. Each job runs in an independent [executor](#executors-and-images) (container or virtual machine). [Caches](#cache) are available to optimize and speed up pipelines, and [workspaces/artifacts](#workspaces-and-artifacts) can be used to share data across your pipeline.

Reviewer: Nice summary.

Author: Thanks!

     # and docker-compose 1.14.0, follow CircleCI Discuss Announcements
     # for new image releases.
       image: ubuntu-1604:201903-01
#...       
   build3:
     macos: # Specifies a macOS virtual machine with Xcode version 11.3
       xcode: "11.3.0"
# ...          
 ```

The Primary Container is defined by the first image listed in [`.circleci/config.yml`]({{ site.baseurl }}/2.0/configuration-reference/) file. This is where commands are executed. The Docker executor spins up a container with a Docker image. The machine executor spins up a complete Ubuntu virtual machine image. See [Choosing an Executor Type]({{ site.baseurl }}/2.0/executor-types/) document for a comparison table and considerations. Subsequent images can be added to spin up Secondary/Service Containers.

When using the docker executor and running docker commands, the `setup_remote_docker` key can be used to spin up another docker container in which to run these commands, for added security. For more information see the [Running Docker Commands]({{ site.baseurl }}/2.0/building-docker-images/#accessing-the-remote-docker-environment) guide.

#### Steps
{:.no_toc}

> Steps are actions that need to be taken to complete your job. Steps are usually a collection of executable commands. For example, the [`checkout`]({{ site.baseurl }}//2.0/configuration-reference/#checkout) step checks out the source code for a job over SSH. Then, the `run` step executes the `make test` command using a non-login shell by default. Commands can also be defined [outside the job declaration]({{ site.baseurl }}/2.0/configuration-reference/#commands-requires-version-21), making them reusable across your config.

Reviewer: ```suggestion
Steps are actions that need to be taken to complete your job. Steps are usually a collection of executable commands. For example, the [`checkout`]({{ site.baseurl }}//2.0/configuration-reference/#checkout) step, which is a _built-in_ step available across all CircleCI projects, checks out the source code for a job over SSH. Then, the `run` step allows you to run custom commands, such as executing the command `make test`  using a non-login shell by default. Commands can also be defined [outside the job declaration]({{ site.baseurl }}/2.0/configuration-reference/#commands-requires-version-21), making them reusable across your config.
```

Just a suggestion to differentiate between reserved/built-in steps and a user's custom steps. I'm not sure if the "make test" is referring to anything specific, so my suggestion tries to make clear that it is a custom command.

Author: Yes this is definitely clearer now, thank you!!

```
{% endraw %}

### Caches, Workspaces and Artifacts

![workflow illustration]( {{ site.baseurl }}/assets/img/docs/workspaces.png)

A cache stores a file or directory of files such as dependencies or source code in object storage.
Each job may contain special steps for caching dependencies from previous jobs to speed up the build.

{% raw %}

```yaml
version: 2
jobs:
  build1:
    docker: # Each job requires specifying an executor
    # (either docker, macos, or machine), see
    # circleci.com/docs/2.0/executor-types/ for a comparison
    # and more examples.
      - image: circleci/ruby:2.4-node
      - image: circleci/postgres:9.4.12-alpine
    steps:
      - checkout
      - save_cache: # Caches dependencies with a cache key
      # template for an environment variable,
      # see circleci.com/docs/2.0/caching/
          key: v1-repo-{{ .Environment.CIRCLE_SHA1 }}
          paths:
            - ~/circleci-demo-workflows

  build2:
    docker:
      - image: circleci/ruby:2.4-node
      - image: circleci/postgres:9.4.12-alpine
    steps:
      - restore_cache: # Restores the cached dependency.
          key: v1-repo-{{ .Environment.CIRCLE_SHA1 }}       
```

{% endraw %}

> Workspaces are a workflows-aware storage mechanism. A workspace stores data unique to the job, which may be needed in downstream jobs. Each workflow has a temporary workspace associated with it. The workspace can be used to pass along unique data built during a job to other jobs in the same workflow.

Reviewer: ```suggestion
Workspaces are a workflow-aware storage mechanism. A workspace stores data unique to the job, which may be needed in downstream jobs. Each workflow has a temporary workspace associated with it. The workspace can be used to pass along unique data built during a job to other jobs in the same workflow.
```
